docs/design/event-handler-framework.md

```diff
@@ -0,0 +1,223 @@
+type SchedulerCache struct {
+   ......
+   CustomInformers     []cache.SharedIndexInformer
+   ......
+}
+```
+
+Start the custom informer when `SchedulerCache` starts
+
+```go
+// Run  starts the schedulerCache
+func (sc *SchedulerCache) Run(stopCh <-chan struct{}) {
+   go sc.podInformer.Informer().Run(stopCh)
+   ......
+   for _, i := range sc.CustomInformers {
+      go i.Run(stopCh)
+   }
+   ......
+}
+```
+
+Wait the custom informer to synced before the starting of scheduling loop
+
+```go
+// WaitForCacheSync sync the cache with the api server
+func (sc *SchedulerCache) WaitForCacheSync(stopCh <-chan struct{}) bool {
+   return cache.WaitForCacheSync(stopCh,
+      func() []cache.InformerSynced {
+         informerSynced := []cache.InformerSynced{
+            sc.podInformer.Informer().HasSynced,
+            ......
+         }
+         ......
+      }()...,
+   ) && func() bool {
+         for _, i := range sc.CustomInformers {
+            if !i.HasSynced() {
+               return false
+            }
+         }
+         
+         return true
+   }
+}
+```
+
+### EventHandler Factory
+
+- define an interface, including object event processing function set
+- provide a factory to realize the object event processing function
+- provide a factory registration mechanism
+
+```go
+// Interface containers all involved object event processing interfaces.
+type Interface interface {
+   AddPod(obj interface{})
+   // other object event processing interfaces.
+}
+
+type Factory func(cache *SchedulerCache, config io.Reader, restConfig *rest.Config) (Interface, error)
+
+var eventHandlers = make(map[string]Factory)
+var eventHandlerMutex sync.Mutex
+
+func RegisterEventHandler(name string, eventHandler Factory) {
+   eventHandlerMutex.Lock()
+   defer eventHandlerMutex.Unlock()
+   if _, found := eventHandlers[name]; found {
+      klog.Fatalf("EventHandler %q was registered twice.", name)
+   }
+   klog.V(1).Infof("Registered eventHandler %q.", name)
+   eventHandlers[name] = eventHandler
+}
+
+func GetEventHandler(name, configFilePath string, cache *SchedulerCache, restConfig *rest.Config) (Interface, error) {
+   eventHandlerMutex.Lock()
+   defer eventHandlerMutex.Unlock()
+   f, found := eventHandlers[name]
+   if !found {
+      return nil, nil
+   }
+   
+   var config *os.File
+   if len(configFilePath) != 0 {
+      config, err = os.Open(configFilePath) 
+   }
+   return f(cache, config, restConfig), nil
+}
+```
+
+#### Implement the default eventHandler
+
+```go
+type defaultEventHandler struct{
+   cache *SchedulerCache
+}
+
+func init() {
+   RegisterEventHandler("default", func(cache *SchedulerCache, config io.Reader, restConfig *rest.Config) (i Interface, err error) {
+      return &defaultEventHandler{cache:cache}, nil
+   })
+}
+
+func (eh *defaultEventHandler) AddPod(obj interface{}) {
+   pod, ok := obj.(*v1.Pod)
+   if !ok {
+      klog.Errorf("Cannot convert to *v1.Pod: %v", obj)
+      return
+   }
+
+   sc := eh.cache
```

Comment:
Is there any change against the package structure?  new added eventhandler and schedulerCache are still in the same package?

Reply:
The default eventhandler is implemented in the original package.